### PR TITLE
Mesh filter bugfix and other minor changes

### PR DESCRIPTION
**Pull request by paulromano**
This PR makes a number of small changes/fixes in the Python API. The only serious bug fix is related to our MeshFilter object. When working on an implementation of a mesh surface filter (to be used for CMFD), I noticed that there is a bug with indexing in our Python implementation of mesh filters. In particular, it assumes that when you have a 3D mesh, the lowest stride in memory appears for the z-axis and the largest stride appears for the x-axis. On the Fortran side, it is clear that it is actually the exact opposite: the x-axis has the lowest stride. This becomes evident if you compare results from tallies.out with what you get from a Tally dataframe. I've fixed the indexing so that they should match now. This does result in a few test results changing.

The other minor changes include:

   - More use of the `uncertainties` package: for example, `k_combined` is now represented as a number with uncertainty (ufloat), as are results of stochastic volume calculations. I haven't gone all the way to using it for tally results yet as that would be a much bigger change.
   - Support for energyout filters in the C API
   - Changed the return values for a few functions in `openmc.data`. Basically, where functions could have returned a value or None, I've changed them to only return a value and otherwise raise exceptions. This is considered more Pythonic.
   - The default density units on a `Material` is now 'sum'
   - When adding a nuclide/element to a material, if the name looks weird or inconsistent with GND, the user will receive a warning.
   - When adding actinides to a material, it is automatically marked as depletable unless otherwise indicated.
   - Added a `Plot.to_image` function to address #974. `Universe.plot` also returns a proper image object now as well.
   - Once again made matplotlib imports optional. I've run into cases trying to use OpenMC on supercomputers where importing matplotlib causes issues.
   - `StatePoint.date_and_time` is actually returned as a `datetime` object.
   - When performing tally arithmetic, numpy will now ignore divide-by-zero warnings.
   - Use 'latest' HDF5 format when building JEFF library (a user ran into an error because of the earliest format)

### CODE REVIEW

File: openmc/data/data.py [hunk 1]
```diff
@@ -1,10 +1,9 @@
 import itertools
+from math import sqrt
```
Comment (smharper) [inline openmc/data/data.py]: Out of curiosity, why do you prefer `math.sqrt` over `numpy.sqrt`?

File: openmc/filter.py [hunk 2]
```diff
@@ -732,37 +732,40 @@ def get_bin_index(self, filter_bin):
...
             return (x, y, z)
 
-        # Construct 2-tuple of x,y cell indices for a 2D mesh
-        else:
+        elif n_dim == 2:
+            # Construct 2-tuple of x,y cell indices for a 2D mesh
             nx, ny = self.mesh.dimension
-            x = bin_index / ny
-            y = bin_index - (x * ny)
+            x = bin_index % nx + 1
```
Comment (smharper) [inline openmc/filter.py]: Put the modulo terms in parentheses for clarity.

File: openmc/plots.py [hunk 3]
```diff
@@ -650,6 +651,42 @@ def to_xml_element(self):
 
         return element
 
+    def to_image(self, openmc_exec='openmc', cwd='.', convert_exec='convert'):
+        """Render plot as an image
```
Comment (smharper) [inline openmc/plots.py]: I find `to_image` a little vague.  How about `render_ipython_image`?

File: openmc/data/data.py (hunk 1 above)
Comment (paulromano) [reply openmc/data/data.py]: The `sqrt` is only used for scalar quantities. `numpy.sqrt` is appropriate if you're actually going to take the square root of a vector, but if you're only using it for a scalar, it adds overhead compared to `math.sqrt`.

File: openmc/filter.py (hunk 2 above)
Comment (paulromano) [reply openmc/filter.py]: Good idea

File: openmc/plots.py (hunk 3 above)
Comment (paulromano) [reply openmc/plots.py]: I'm trying to stick to the convention of `to_*` meaning "convert this to another type" which we use in many other places in the API (note the difference between that and `export_to_*` which implies that a file is written). I can rename it to `to_ipython_image` though; how about that?

Comment (smharper) [reply openmc/plots.py]: Sure, `to_ipython_image` sounds good.  Although I'd also like you to mention in the docstring that this is calling OpenMC (to make it clear this is a potentially expensive function that requires all settings, geometry, and materials XMLs to already be exported).

**Comment by smharper** [comment]
I made the following changes to the pincell XML example, and I get an "arrays must all be the same length error" when making a dataframe:

```
--- a/examples/xml/pincell/tallies.xml
+++ b/examples/xml/pincell/tallies.xml
@@ -2,13 +2,13 @@
 <tallies>
 
   <mesh id="2" type="regular">
-    <dimension>100 100 1</dimension>
-    <lower_left>-0.62992 -0.62992 -1.e50</lower_left>
-    <upper_right>0.62992  0.62992  1.e50</upper_right>
+    <dimension>2 2 10</dimension>
+    <lower_left>-0.62992 -0.62992 -5</lower_left>
+    <upper_right>0.62992  0.62992  5</upper_right>
   </mesh>
 
   <filter id="1" type="mesh">
-    <bins>1</bins>
+    <bins>2</bins>
   </filter>
 
   <filter id="2" type="energy">
@@ -20,4 +20,9 @@
     <scores>flux fission nu-fission</scores>
   </tally>
 
+  <tally id="2">
+    <filters>1</filters>
+    <scores>current</scores>
+  </tally>
+
 </tallies>
```

**Comment by paulromano** [comment]
Your filter would have to be `type="meshsurface"` in order to work with a current score. I'll add in an input check to make sure a user doesn't try to do a current score with a normal mesh filter.

**Comment by smharper** [comment]
Whoops.  While you're at it, change the filter bin to "2" in this example.

**Comment by smharper** [comment]
Also, it looks like OpenMC segfaults if you give it a meshsurface filter on a tally like the one with id=1 in the pincell example.

**Comment by paulromano** [comment]
Sorry, getting my branches confused. I have another branch (building off of this one) with a proper mesh surface filter implemented. I'm not sure if dataframes ever worked with mesh filters used for currents. My recommendation would be to ignore the mesh current issues for now since I will be submitting another PR dealing with after we're done with this one.

**Comment by smharper** [comment]
Yeah, I don't think our dataframe mesh currents have ever been correct either.  Sure, I'll merge now and await further repairs down the line.

**Pull request merged**